### PR TITLE
typeck: refactor patterns => `pat.rs`  + make the `def_bm` algo more declarative

### PR DESCRIPTION
**Pull request by Centril**
Spurred by the relative difficulty I had in working up an explanation of how default match bindings work in https://github.com/rust-lang/rust/pull/63118#issuecomment-524161584, this PR refactors the type checking of patterns into `pat.rs`.

The PR is probably best read commit-by-commit and includes various changes beyond the following, which are the most important highlights:

- The algorithm for determining `def_bm` is encoded in a more declarative fashion now with important sub-steps divided into functions that make sense as logical units (and as described in the reference). This is done starting with *"extract `is_no_ref_pat`."* to *"extract `calc_default_binding_mode`"*.

- Dedicated functions like `check_pat_{lit,range,ident,tuple,box,ref,slice}` are then introduced for the various kinds of patterns to make things overall more readable.

- `fn check_pat_top(...)` becomes the sole entry point to type checking patterns.

   This will take care of initializing the default binding mode (hence: `def_bm`) to `BindByValue` and is called by all contexts that have a pattern that needs to be type checked (functions, `match`, `if let`, `let`, ...). The overall result is that the notion of `def_bm` is internal to checking patterns.

- Various diagnostics are extracted to dedicated functions to disturb the flow of type checking logic less.

r? @oli-obk

### CODE REVIEW

File: src/librustc_typeck/check/pat.rs [hunk 1]
```diff
@@ -1059,4 +1016,58 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        err.emit();
+                    }
+                    (rptr_ty, inner_ty)
+                }
+            };
+
+            self.check_pat_walk(&inner, inner_ty, def_bm, discrim_span);
+            rptr_ty
+        } else {
+            self.check_pat_walk(&inner, tcx.types.err, def_bm, discrim_span);
```
Comment (oli-obk) [inline src/librustc_typeck/check/pat.rs]: Dedup this call between the arms

Comment (Centril) [reply src/librustc_typeck/check/pat.rs]: Nicely spotted! (and done)

**Comment by Centril** [comment]
@oli-obk I've done a similar dedup in `check_pat_box` also. =)

**Review by oli-obk — APPROVED** [review]
@bors r+ p=1

**Comment by Centril** [comment]
Let's try this again... ;)

@bors r=oli-obk p=1

**Pull request merged**